src/lib/util/print.c

```diff
@@ -838,3 +838,38 @@ char *fr_asprintf(TALLOC_CTX *ctx, char const *fmt, ...)
 
 	return ret;
 }
+
+/** Special version of fprintf which implements custom format specifiers
+ *
+ * @copybrief fr_vasprintf
+ *
+ * @param[in] ctx	to allocate buffer in.
+ * @param[in] fp	to write the result of fmt string.
+ * @param[in] fmt	string.
+ * @param[in] ...	variadic argument list.
+ * @return
+ *   - On success, the number of bytes written is returned (zero indicates nothing was written).
+ *   - On error, -1 is returned, and errno is set appropriately
+ */
+int fr_fprintf(TALLOC_CTX *ctx, FILE *fp, char const *fmt, ...)
```

Comment:
This doesn't need to take a talloc ctx.  If you want it to be efficient, you should allocate a thread specific memory pool and use that instead... But I think fr_vasprintf already does that for the intermediary buffers, so you won't save  that much CPU time.